ci_feature/main.sh

```diff
@@ -186,26 +164,153 @@ echo "Making wget request to cadvisor endpoint with port 10250"
+export CONTAINER_RUNTIME="docker"
+export NODE_NAME=""
+
 if [ "$cAdvisorIsSecure" = true ] ; then
       echo "Wget request using port 10250 succeeded. Using 10250"
       export IS_SECURE_CADVISOR_PORT=true
       echo "export IS_SECURE_CADVISOR_PORT=true" >> ~/.bashrc
       export CADVISOR_METRICS_URL="https://$NODE_IP:10250/metrics"
       echo "export CADVISOR_METRICS_URL=https://$NODE_IP:10250/metrics" >> ~/.bashrc
+      echo "Making wget request to cadvisor endpoint /pods with port 10250 to get the configured container runtime on kubelet"
+      IS_SUCCESS=$(wget --server-response https://$NODE_IP:10250/pods --no-check-certificate --header="Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" -O podsResponseFile 2>&1 | grep -c '200 OK')      
+      if [ $IS_SUCCESS == 1 ]; then    
+         podsResponse=$(cat podsResponseFile)        
+         ITEMS_COUNT=$(echo $podsResponse | jq '.items | length')
+         echo "found items count: $ITEMS_COUNT"
+         if [ $ITEMS_COUNT -gt 0 ]; then 
+            # exclude the pods which doesnt have containerId. could happen if the container fails to start because of bad image and tag etc..
+            podsWithValidContainerId=$(echo $podsResponse | jq -r '[.items[] | select( .status.containerStatuses != null and .status.containerStatuses[].containerID != null and  .status.containerStatuses[].containerID != "")]')                         
+            ITEMS_COUNT_WITH_CONTAINER_ID=$(echo $podsWithValidContainerId | jq '. | length')
+            if [ $ITEMS_COUNT_WITH_CONTAINER_ID -gt 0 ]; then 
+                  # containerId name contract https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#pod-v1-core
+                  containerRuntime=$(echo $podsWithValidContainerId | jq -r '.[0].status.containerStatuses[0].containerID' | cut -d ':' -f 1)                  
+                  nodeName=$(echo $podsWithValidContainerId | jq -r '.[0].spec.nodeName')
+                  # convert to lower case so that everywhere else can be used in lowercase
+                  containerRuntime=$(echo $containerRuntime | tr "[:upper:]" "[:lower:]")
+                  nodeName=$(echo $nodeName | tr "[:upper:]" "[:lower:]")
+                  # update runtime only if its not empty and not startswith docker
+                  if [ -z $containerRuntime ]; then
+                      echo "using default container runtime as docker since got containeRuntime as empty string"                              
+                  elif [[ $containerRuntime != docker* ]]; then                     
+                       export CONTAINER_RUNTIME=$containerRuntime                                          
+                  fi
+
+                  if [ -z $nodeName ]; then
+                      echo "-e error nodeName in /pods API response is empty"
+                  else
+                     export NODE_NAME=$nodeName   
```

Comment:
I see that you are using NODE_NAME to set the containerhostname, from this logic what if the first pod we get is in pending state?

Reply:
> I see that you are using NODE_NAME to set the containerhostname, from this logic what if the first pod we get is in pending state?

> all the bad pods are filtered out before reaching here.